### PR TITLE
chore: eth69 status message support

### LINKED ISSUE #16073: Update eth69 status message
### Describe the feature

with https://github.com/ethereum/devp2p/pull/262/files

we need to

a) update the fields:

https://github.com/paradigmxyz/reth/blob/f24570844545b55f2293a3a3c475133a3f738b68/crates/net/eth-wire-types/src/status.rs#L233-L233

and because this will require a set of hashes, we finally need to update this update function

https://github.com/paradigmxyz/reth/blob/f24570844545b55f2293a3a3c475133a3f738b68/crates/net/network/src/network.rs#L414-L417

which is only used for:

https://github.com/paradigmxyz/reth/blob/f24570844545b55f2293a3a3c475133a3f738b68/crates/net/network/src/session/mod.rs#L217-L221

- `earliest`: number of the earliest available full block
- `latest`: number of the latest available full block number
- `latestHash`: hash of the latest available full block

The way this works is that `SessionManager` tracks the `Status`:

https://github.com/paradigmxyz/reth/blob/f24570844545b55f2293a3a3c475133a3f738b68/crates/net/network/src/session/mod.rs#L79-L80

now 69 requires some extra info that we'd also need to track and need to push via the `update_status` or other additional functions

and there's actually more to do because this currently wouldn't quire work for eth69

https://github.com/paradigmxyz/reth/blob/f24570844545b55f2293a3a3c475133a3f738b68/crates/net/eth-wire/src/handshake.rs#L81-L95

invoked here:

https://github.com/paradigmxyz/reth/blob/f24570844545b55f2293a3a3c475133a3f738b68/crates/net/network/src/session/mod.rs#L1059-L1068

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shane-moore**
## Changes
- resolves https://github.com/paradigmxyz/reth/issues/13186?issue=paradigmxyz%7Creth%7C16073
- additionally, `SessionManager`, `SessionEvent`, `PeerInfo` etc have their nested prop `status: Status` updated to `status: StatusMessage`

where
```
pub enum StatusMessage {
    /// The legacy status (`eth/66` through `eth/68`) with `total_difficulty`.
    Legacy(Status),
    /// The new `eth/69` status with no `total_difficulty`.
    Eth69(StatusEth69),
}
```

Now, `eth_handshake` no longer has the dependency of legacy status as a param or return value. These have been updated to `StatusMessage` as well. Should be simpler now since no longer need to convert a potential eth69 status back to a legacy status in `eth_handshake`.  It's also nice that `SessionManager` no longer only supports legacy status

## PR Status
- code compiles

## Todo:
- fix tests
- resolve comments below

@mattsse, lmk if we like this approach, and I'll finish it up

### CODE REVIEW

File: crates/net/network/src/config.rs [hunk 1]
```diff
@@ -622,7 +623,9 @@ impl<N: NetworkPrimitives> NetworkConfigBuilder<N> {
         hello_message.port = listener_addr.port();
 
         // set the status
-        let status = Status::spec_builder(&chain_spec, &head).build();
+        // eth69 todo: think about if we want to support eth69 status message when building
+        // NetworkConfig
+        let status = StatusMessage::Legacy(Status::spec_builder(&chain_spec, &head).build());
```
Comment (shane-moore) [inline crates/net/network/src/config.rs:628]: ideally, we'd use the `EthVersion` we're running to determine whether `status` should be `Legacy` or `Status69` variant, but it looks like we don't define the version til later https://github.com/paradigmxyz/reth/blob/5c64d5915332593d45ad96f99a7bc395102e2e4a/crates/net/network/src/session/mod.rs#L1046-L1068

we could perhaps default to `Legacy` variant in the `NetworkConfig`. then, use a builder to update `status` to `StatusMessage:Eth69` if the version is `69`, right before the `match handshake`. open to suggestions on this

File: crates/net/network/src/network.rs [hunk 2]
```diff
@@ -413,7 +418,12 @@ impl<N: NetworkPrimitives> NetworkSyncUpdater for NetworkHandle<N> {
 
     /// Update the status of the node.
     fn update_status(&self, head: Head) {
-        self.send_message(NetworkHandleMessage::StatusUpdate { head });
+        // Derive earliest_block as the Paris (Merge) activation block:
+        let merge_block = ChainSpec::default().paris_block().unwrap_or(0);
+        let earliest = if head.number >= merge_block { U64::from(merge_block) } else { U64::ZERO };
```
Comment (shane-moore) [inline crates/net/network/src/network.rs:423]: I set earliest to the pre-merge block since that was called out in the eip https://eips.ethereum.org/EIPS/eip-7642#:~:text=In%20the%20history%20expiry%20working%20group%2C%20it%20was%20decided%20that%20clients%20may%20drop%20pre%2Dmerge%20history%20from%20their%20storage%20after%20May%201

lmk if another implementation is preferable

File: crates/net/eth-wire-types/src/status.rs [hunk 3]
```diff
@@ -74,9 +74,11 @@ impl Status {
         StatusEth69 {
             version: EthVersion::Eth69,
             chain: self.chain,
-            blockhash: self.blockhash,
             genesis: self.genesis,
             forkid: self.forkid,
+            earliest: U64::ZERO, // Will be set from actual chain state in network layer
+            latest: U64::ZERO,   // Will be set from actual chain state in network layer
```
Comment (mattsse) [inline crates/net/eth-wire-types/src/status.rs:80]: this makes sense

File: crates/net/eth-wire-types/src/status.rs [hunk 4]
```diff
@@ -251,9 +251,26 @@ pub struct StatusEth69 {
     /// [EIP-2124](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2124.md).
     /// This was added in [`eth/64`](https://eips.ethereum.org/EIPS/eip-2364)
     pub forkid: ForkId,
+
+    /// Earliest block number this node can serve
+    pub earliest: U64,
+
+    /// Latest block number this node has (current head)
+    pub latest: U64,
```
Comment (mattsse) [inline crates/net/eth-wire-types/src/status.rs:259]: I'd like to keep these as primitive u64

File: crates/net/eth-wire-types/src/status.rs [hunk 5]
```diff
@@ -377,6 +410,14 @@ impl Encodable for StatusMessage {
     }
 }
 
+impl Display for StatusMessage {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            StatusMessage::Legacy(s) => Display::fmt(s, f),
+            StatusMessage::Eth69(s69) => Display::fmt(s69, f),
```
Comment (mattsse) [inline crates/net/eth-wire-types/src/status.rs:417]: you can write this as `s.fmt(f)`

File: crates/net/network/src/config.rs [hunk 6]
```diff
@@ -73,7 +74,7 @@ pub struct NetworkConfig<C, N: NetworkPrimitives = EthNetworkPrimitives> {
     /// The executor to use for spawning tasks.
     pub executor: Box<dyn TaskSpawner>,
     /// The `Status` message to send to peers at the beginning.
-    pub status: Status,
+    pub status: StatusMessage,
```
Comment (mattsse) [inline crates/net/network/src/config.rs:77]: this is a bit problematic because we can't convert this in both directions, for example if we configure 69, we can really convert this fully to <=68

so I think we need to keep Status here and in addition to that we need to a way to configure the block range for 69?
which we eventually need to set externally if there's history expiry.

so we can introduce two more fields here for the block range?

File: crates/net/network/src/network.rs [hunk 7]
```diff
@@ -108,7 +109,11 @@ impl<N: NetworkPrimitives> NetworkHandle<N> {
 
     /// Update the status of the node.
     pub fn update_status(&self, head: Head) {
-        self.send_message(NetworkHandleMessage::StatusUpdate { head });
+        let merge_block = ChainSpec::default().paris_block().unwrap_or(0);
+        let earliest = if head.number >= merge_block { U64::from(merge_block) } else { U64::ZERO };
+        let latest = U64::from(head.number);
+
+        self.send_message(NetworkHandleMessage::StatusUpdate { head, earliest, latest });
```
Comment (mattsse) [inline crates/net/network/src/network.rs:116]: after looking at this, I think we can keep this unchanged for now, if we ignore the case for now where the earliest block can change, so we dont actually need to change this

File: crates/net/network/src/session/mod.rs [hunk 8]
```diff
@@ -77,7 +78,7 @@ pub struct SessionManager<N: NetworkPrimitives> {
     /// The secret key used for authenticating sessions.
     secret_key: SecretKey,
     /// The `Status` message to send to peers.
-    status: Status,
+    status: StatusMessage,
```
Comment (mattsse) [inline crates/net/network/src/session/mod.rs:81]: same problem as with the NetworkConfig,

we can also introduce the two additional fields, for earliest,latest in additon,

maybe we should introduce a new super type for all status fields, which is a superset of status and status69?

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, this is a great start,

I left some suggestions, and after looking at this a bit closer we can treat the changes to StatusUpdate as out of scope for this pr, because from head we also get the latest block.

File: crates/net/network/src/config.rs (hunk 6 above)
Comment (shane-moore) [reply crates/net/network/src/config.rs]: I feel you on the loss of fidelity if try to bi-directionally convert. I updated this to use a `UnifiedStatus`, which I'll elaborate on a couple comments down

Comment (shane-moore) [reply crates/net/network/src/config.rs]: and for being able to externally set the block range, we should be able to use `UnifiedStatus::set_history_range`

File: crates/net/network/src/session/mod.rs (hunk 8 above)
Comment (shane-moore) [reply crates/net/network/src/session/mod.rs]: in latest commit https://github.com/paradigmxyz/reth/commit/8b1f040f3356cc766cba3a6f7463fd1b3923d7b0, I went with the unified status and status69 approach. The concept is that we store the superset of field in UnifiedStatus for use locally. and then, only use StatusMessage for rlp encoding and sending in eth_handshake. I went with this approach because it seems more extensible overtime. should be as easy as adding new fields to UnifiedStatus for future wire protocol version. additionally, nice not to have to track additional fields latest and earliest

@mattsse, lmk if like this approach or if prefer to keep things as legacy Status + two additional fields. if should proceed, I'll start on test refactoring and additionally see if I can uncover any edge cases

File: crates/net/eth-wire-types/src/status.rs [hunk 9]
```diff
@@ -377,6 +393,66 @@ impl Encodable for StatusMessage {
...
+        }
+    }
+}
+
+impl From<UnifiedStatus> for StatusMessage {
+    fn from(s: UnifiedStatus) -> Self {
+        if s.version == EthVersion::Eth69 {
```
Comment (mattsse) [inline crates/net/eth-wire-types/src/status.rs:407]: can we move the impls of these from< impls into functions of UnifiedStatus (`into_legacy` + `into_status69`) this makes it easier to find these and obvious which conversion is called

File: crates/net/network/src/config.rs [hunk 10]
```diff
@@ -73,7 +73,7 @@ pub struct NetworkConfig<C, N: NetworkPrimitives = EthNetworkPrimitives> {
     /// The executor to use for spawning tasks.
     pub executor: Box<dyn TaskSpawner>,
     /// The `Status` message to send to peers at the beginning.
-    pub status: Status,
+    pub status: UnifiedStatus,
```
Comment (mattsse) [inline crates/net/network/src/config.rs:76]: great, exactly how this should be solved

**Review by mattsse — CHANGES_REQUESTED** [review]
okay, sorry about the delay, wanted to get the next release out first.

this is looking great, I only have a few last nits

File: crates/net/eth-wire-types/src/status.rs (hunk 9 above)
Comment (shane-moore) [reply crates/net/eth-wire-types/src/status.rs]: @mattsse, I feel you on this. nice to have all the conversion logic encapsulated in the `UnifiedStatus` impl. updated in latest commit https://github.com/paradigmxyz/reth/pull/16099/commits/19828ede5bc0c95108e27bcf0a9d69db5f407fc2

will start on the tests tomorrow 🤠

File: crates/net/eth-wire/src/ethstream.rs [hunk 11]
```diff
@@ -87,20 +87,21 @@ where
...
-        EthereumEthHandshake(&mut self.inner).eth_handshake(status, fork_filter).await?;
+        let their_status =
+            EthereumEthHandshake(&mut self.inner).eth_handshake(status, fork_filter).await?;
 
         // now we can create the `EthStream` because the peer has successfully completed
         // the handshake
         let stream = EthStream::new(status.version, self.inner);
 
-        Ok((stream, status))
+        Ok((stream, their_status))
```
[the remark this answers is not in the document]
Comment (shane-moore) [reply crates/net/eth-wire/src/ethstream.rs:104]: It seems arbitrary whether we use `their_status.version` or `status.version` since we have already set our version to match their's before getting to this handshake step, as shown in https://github.com/paradigmxyz/reth/blob/a62bde37cae4f446cfb3edf718911ae4e5942633/crates/net/network/src/session/mod.rs#L1058-L1072

def open to making this modification if preferable though

Comment (shane-moore) [inline crates/net/eth-wire/src/ethstream.rs:104]: @mattsse, the previous implementation of `handshake_without_timeout` was returning the local status instead of `their_status` in the tuple.  This seems incorrect since downstream functions expect to receive a `their_status` during the handshake flow, so I updated it here.

File: crates/net/eth-wire-types/src/status.rs (hunk 9 above)
Comment (shane-moore) [reply crates/net/eth-wire-types/src/status.rs]: @mattsse, updated the test coverage to support `UnifiedStatus` + resolved linting in latest commit https://github.com/paradigmxyz/reth/pull/16099/commits/edd85b3ecfcb57288a2afd348539f5eb1c14e7c6

File: crates/net/eth-wire/src/ethstream.rs (hunk 11 above)
Comment (mattsse) [reply crates/net/eth-wire/src/ethstream.rs:104]: yeah this makes sense

**Review by mattsse — APPROVED** [review]
nice, ty!
